### DISCUSSION TITLE
Get a `pyclass` from a `pyclass` with reference?

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#2261] [Questions] [answered]
I have two `#[pyclass]` and I combine one into another one like below:

```rust
#[pyclass]
struct Wallet {
    pub cash: i64,
}

#[pymethods]
impl Wallet {
    #[new]
    fn new(cash: i64) -> Self {
        Self { cash }
    }
    pub fn add_cash(&mut self, cash: i64) {
        self.cash += cash;
    }
}

#[pyclass]
struct Player {
    pub name: String,
    #[pyo3(get)]
    pub wallet: Wallet,
    #it seems using #[gettar] will lost `self.wallet` ?
}
#[pymethods]
impl Player {
    #[new]
    fn new(name: String, cash: i64) -> Self {
        Self {
            name,
            wallet: Wallet { cash },
        }
    }
}
```
What I want to do is create a instance of `Player` in Python and call methods of `Wallet` like: 
```python
player = Player("haha", 999)
player.wallet.add_cash(10)
```
But I found everytime use `player.wallet`, it would return a new clone for wallet with a new memory address(found by call `id(wallet)`). 
And because `player.wallet` need to create new object, it takes more time.

How to achieve `player.wallet.add_cash(10)` in Python like Rust did (call its inner instance rather than a clone) ?

**Comment by adamreichold** [comment, marked as answer]
```rust

#[pyclass]
struct Wallet {
    pub cash: i64,
}

#[pymethods]
impl Wallet {
    #[new]
    fn new(cash: i64) -> Self {
        Self { cash }
    }
    pub fn add_cash(&mut self, cash: i64) {
        self.cash += cash;
    }
}

#[pyclass]
struct Player {
    pub name: String,
    #[pyo3(get)]
    pub wallet: Py<Wallet>,
}

#[pymethods]
impl Player {
    #[new]
    fn new(name: String, cash: i64, py: Python) -> Self {
        Self {
            name,
            wallet: Py::new(py, Wallet { cash }).unwrap(),
        }
    }
}
```

**Reply by sun-rs** [reply]
Thanks for reply. So is this answer means that I can't use `player.acc.method()` in python while remain code not changed in Rust(because I also need to use in pure Rust)? Must change `Wallet` to `Py<Wallet>` to get a reference in Python?

**Reply by adamreichold** [reply]
>  while remain code not changed in Rust(because I also need to use in pure Rust)?

Yes, but calling a method like `self.wallet.borrow_mut(py).add_cash(42);` is also "pure" Rust. The borrow check overhead _is required_ because by design the `Wallet` instance is shared with and mutated by the Python code.

**Reply by sun-rs** [reply]
> > while remain code not changed in Rust(because I also need to use in pure Rust)?
> 
> Yes, but calling a method like `self.wallet.borrow_mut(py).add_cash(42);` is also "pure" Rust. The borrow check overhead _is required_ because by design the `Wallet` instance is shared with and mutated by the Python code.

When call `Py<Wallet>` in pure Rust(without any python action), I know it must be slower. But if write a wrapper method in `Player` for `add_cash` and call `Player.add_cash` in Python, is this a better way for execute speed while remain the member of struct `Player` not change?

**Reply by adamreichold** [reply]
>  is this a better way for execute speed while remain the member of struct Player not change?

Please measure it. Whether there is an increase in efficiency and whether it is worth the reduced functionality cannot be answered in general.
